src/hooks.c

```diff
@@ -0,0 +1,203 @@
+        sizeof(struct hptr));
+
+    return;
+}
+
+void hooks_deregister(void)
+{
+    free(hooks_available);
+    hooks_available = NULL;
+    return;
+}
+
+inline bool hooklist_run(Hooklist h, Message msg)
+{
+    bool res;
+    for (uint64_t i = 0; i < h->num; i++)
+    {
+        Hptr hook = *(h->hptrarray+i);
+
+        res = hook->hook(hook->ctx,msg);
+        if(!res){
+            // free message (unusable)
+            free(message_get_data(msg));
+            message_set_data(msg, NULL);
+            message_set_len(msg,0);
+            metadata_free(message_get_metadata(msg));
+
+            return false;
+        }
+    }
+    return true;
+}
+
+bool hooks_validate(config_setting_t *conf)
+{
+    bool res = true;
+
+    size_t list;
+    config_setting_t *hook = NULL, *type = NULL;
+
+    if(!config_setting_is_list(conf)) {
+        res = false;
+        goto error;
+    }
+
+    list = config_setting_length(conf);
+
+    for (size_t i = 0; i < list; ++i)
+    {
+        Hptr hookptr = NULL;
+        hook = config_setting_get_elem(conf, i);
+        /* if the list elemnt is undefined, we have walked into
+         * a libconfig bug. */
+        if(hook == NULL)
+            abort();
+
+        if(!(type = CONF_GET_MEM(hook, "type", "hooks need a type!")))
+        {
+            res = false;
+            goto next;
+        }
+
+        const char *name = config_setting_get_string(type);
+        if(name == NULL)
+        {
+            fprintf(stderr, "%s %d: hook is not of type string!\n",
+                __FILE__, __LINE__);
+            res = false;
+            goto next;
+        }
+
+        hookptr = _find_hook(name);
+        if(hookptr == NULL)
+        {
+            fprintf(stderr, "%s %d: %s not a valid hook type!\n",
+                __FILE__, __LINE__, name);
+            res = false;
+            goto next;
+        }
+
+        res &= hookptr->validate(hook);
+
+        next:
+        free(hookptr);
+    }
+
+    error:
+    return res;
+}
+
+void hooks_add(Hooklist h, config_setting_t *conf)
+{
+    size_t list;
+    config_setting_t *hook = NULL, *type = NULL;
+    Hptr hookptr;
+
+    list = config_setting_length(conf);
+
+    for (size_t i = 0; i < list; ++i)
+    {
+        hook = config_setting_get_elem(conf, i);
+
+        if(!(type = CONF_GET_MEM(hook, "type", "hooks need a type!")))
+            abort();
+
+        hookptr = _find_hook(config_setting_get_string(type));
+
+        size_t size = (h->num+1) * sizeof(hookptr);
+        h->hptrarray = realloc(h->hptrarray,size);
+
+        if (h->hptrarray == NULL)
+            abort();
+        hookptr->ctx = hookptr->init(hook);
+
+        *(h->hptrarray+h->num) = hookptr;
+
+        h->num++;
+    }
+    return;
+}
+
+Hooklist hook_init()
+{
+    Hooklist hooks;
+    hooks = SCALLOC(1,sizeof(*hooks));
+    return hooks;
+}
+
+void hook_free(Hooklist h)
+{
+    Hptr *hooks = h->hptrarray;
+    Hptr hook;
+    if(!h)
```

Comment:
Not sure if `h` can be NULL, but it isn't safe to check `h` after getting `h->hptrarray`.

Reply:
Good point.